manager/controlapi/service.go

```diff
@@ -306,6 +306,7 @@ func validateTaskSpec(taskSpec api.TaskSpec) error {
 func validateEndpointSpec(epSpec *api.EndpointSpec) error {
 	// Endpoint spec is optional
 	if epSpec == nil {
+		epSpec = &api.EndpointSpec{Mode: api.ResolutionModeVirtualIP}
```

Comment:
I would suggest doing this outside this function. This function is for validation purposes only and should not have any side effects.

Reply:
I strongly agree. We should not modify the user's spec after it's been sent to Swarmkit. There are Good Reasons to do this.

Reply:
moved it outside

Reply:
also probably return an InvalidArgument here.

Reply:
then we don't need to set any default if we bail out with an error

Reply:
The change essentially is to make EndpointSpec mandatory, correct ? So, to avoid breaking old clients, we're populating it in swarm.

Its OK to return an error from validateEndpointSpec and handle it wherever it is called by filling in an empty EndpointSpec.